source/common/log_sink/conveyor.h

```diff
@@ -0,0 +1,149 @@
+ *
+ * If the producer would like to instead perform an allocation when there are no
+ * "empty" objects available it can do the allocation and move-pass the unique_ptr to
+ * the putLoaded() method.
+ */
+template <typename T> class Conveyor {
+public:
+  /**
+   * Expand the number of objects used in the conveyor by performing
+   * allocations.  Objects will be constructed using the remaining arguments
+   * as constructor params (emplace style), default constructed if only the
+   * count arugment is given.
+   *
+   * @param count The number of objects to allocate and add to the conveyor
+   * @param args Constructor args (variadic).
+   */
+  template <typename... Args> void expand(int count, Args&&... args) {
+    // We do the slower allocation outside of the lock
+    std::vector<std::unique_ptr<T>> allocations(count);
+    for (auto&& p : allocations) {
+      p.reset(new T(std::forward<Args>(args)...));
+    }
+
+    // Then add the newly allocated items into the outbound deque under lock
+    // It doesn't matter which end we add to.
+    std::unique_lock<std::mutex> locked(lock_);
+    for (auto&& p : allocations) {
+      outbound_.push_back(std::move(p));
+    }
+  }
+
+  /**
+   * Contract the number of objects used in the conveyor to free memory.
+   *
+   * If there are not enough objects in the "unloaded" queue the entire unloaded
+   * queue will be emptied.
+   *
+   * @param count The number of objects to try freeing.
+   * @return int The number of objects actually freed.
+   */
+  int contract(int count) {
+    int liberated = 0;
+    while (liberated < count && !outbound_.empty()) {
+      {
+        std::unique_lock<std::mutex> locked(lock_);
+        outbound_.pop_back(); // Implicitly deletes object as reference goes away
+      }
+      liberated++;
+    }
+    return liberated;
+  }
+
+  /**
+   * Try to get an empty T object.
+   *
+   * If no empty objects are available this method will return nullptr.
+   * It will not block.
+   *
+   * @return std::unique_ptr<T> Empty object, or nullptr.
+   */
+  std::unique_ptr<T> getEmpty() {
+    std::unique_lock<std::mutex> locked(lock_);
+    if (outbound_.empty()) {
+      return std::unique_ptr<T>();
+    } else {
+      auto obj = std::move(outbound_.back());
+      outbound_.pop_back();
+      return obj;
+    }
+  }
+
+  /**
+   * Place a loaded object onto the Conveyor for processing.
+   *
+   * @param obj An std::unique_ptr<T> to a loaded object ready for processing
+   */
+  void putLoaded(std::unique_ptr<T>&& obj) {
+    {
+      std::unique_lock<std::mutex> locked(lock_);
+      inbound_.push_front(std::move(obj));
+    }
+    queue_event_.notify_one();
+  }
+
+  /**
+   * Block and wait for one loaded object.
+   *
+   * @return std::unique_ptr<T> A loaded object ready for processing.
+   */
+  std::unique_ptr<T> waitAndGetLoaded() {
+    std::unique_lock<std::mutex> locked(lock_);
+    queue_event_.wait(locked, [this]() -> bool { return !inbound_.empty(); });
```

Comment:
I was thinking the logging and stats thread didn't want to be regular event handling threads to keep things simple and avoid dependencies. But it could be if this is desired by replacing the condition variable with something that uses the event signals feature.  What's the use case for running a full event loop in the logging thread vs. a lightweight thread?